File: frontend/app/src/api/bills.ts
import { apiGet } from './client';
import type { components } from '../shared/contracts/api';

export type BillSummary = {
  bill_id: string;
  title: string;
  congress: number | null;
  bill_type: string | null;
  number: number | null;
  chamber: string | null;
  updated_at?: string | null;
};

export type BillsResponse = {
  bills: BillSummary[];
  total: number;
};

export type BillDetailResponse = components['schemas']['BillDetailResponse'];

export type BillFilters = {
  query?: string;
  congress?: string;
  chamber?: string;
  billType?: string;
};

export function fetchRecentBills(limit = 50, page = 1, filters: BillFilters = {}): Promise<BillsResponse> {
    const params = new URLSearchParams({ limit: String(limit), page: String(page) });
    if (filters.query) params.set('query', filters.query);
    if (filters.congress) params.set('congress', filters.congress);
    if (filters.chamber) params.set('chamber', filters.chamber);
    if (filters.billType) params.set('bill_type', filters.billType);
    return apiGet<BillsResponse>(`/api/v1/bills/recent?${params.toString()}`);
}

export function fetchBill(billId: string): Promise<BillDetailResponse> {
  return apiGet<BillDetailResponse>(`/api/v1/bills/${encodeURIComponent(billId)}`);
}

File: cdm/backend/api/routes/bills.py
from fastapi import APIRouter, Query

from cdm.backend.services.bill_service import get_bill, list_recent_bills
from cdm.backend.services.topic_service import get_bill_topics
from cdm.backend.services.vote_service import list_votes_for_bill
from cdm.contracts.api import (
    BillDetailResponse,
    BillsResponse,
    BillTopicsResponse,
    BillVotesResponse,
)

router = APIRouter(prefix="/bills", tags=["bills"])


@router.get("/recent", response_model=BillsResponse)
def recent_bills(
    limit: int = Query(default=50, ge=10, le=100),
    page: int = Query(default=1, ge=1),
    query: str | None = Query(default=None, max_length=200),
    congress: int | None = Query(default=None, ge=1, le=200),
    bill_type: str | None = Query(default=None, max_length=20),
    chamber: str | None = Query(default=None, pattern="^(House|Senate)$"),
) -> BillsResponse:
    return list_recent_bills(limit, page, query, congress, bill_type, chamber)


@router.get("/{bill_id}", response_model=BillDetailResponse)
def bill_detail(bill_id: str) -> BillDetailResponse:
    return get_bill(bill_id)


@router.get("/{bill_id}/votes", response_model=BillVotesResponse)
def bill_votes(bill_id: str) -> BillVotesResponse:
    return list_votes_for_bill(bill_id)


@router.get("/{bill_id}/topics", response_model=BillTopicsResponse)
def bill_topics(bill_id: str) -> BillTopicsResponse:
    return get_bill_topics(bill_id)
